src/snmalloc/global/memcpy.h

```diff
@@ -179,6 +189,60 @@ namespace snmalloc
     }
   };
 
+  /**
+   * StrictProvenance architectures are prickly about their pointers.  In
+   * particular, they may not permit misaligned loads and stores of
+   * pointer-sized data, even if they can have non-pointers in their
+   * pointer registers.  On the other hand, pointers might be hiding anywhere
+   * they are architecturally permitted!
+   */
+  struct GenericStrictProvenance
+  {
+    static SNMALLOC_FAST_PATH_INLINE void
+    copy(void* dst, const void* src, size_t len)
+    {
+      static_assert(bits::is_pow2(sizeof(void*)));
+
+      // If this is a small size, use a jump table
+      if (len < sizeof(void*))
+      {
+        small_copies<sizeof(void*), sizeof(long)>(dst, src, len);
+      }
+      // Equally-misaligned segments could be holding pointers internally,
+      // so perform unaligned operations at the top and bottom of the range.
+      else if (
+        (address_cast(src) & (sizeof(void*) - 1)) ==
+        (address_cast(dst) & (sizeof(void*) - 1)))
+      {
+        // Come up to pointer-alignment using a jump table
+        unaligned_start<sizeof(void*), sizeof(long)>(dst, src, len);
+
+        // Move pointers for as long as we can (possibly zero times)
+        block_copy<sizeof(void*)>(dst, src, len);
+
+        // There are up to sizeof(void*)-1 bytes left at the end, aligned at
+        // alignof(void*).  Figure out where that is and copy longs for as
+        // long as we can (possibly zero times)
+        dst = pointer_align_down<sizeof(void*)>(pointer_offset(dst, len));
+        src = pointer_align_down<sizeof(void*)>(pointer_offset(src, len));
+        len = len & (sizeof(void*) - 1);
+
+        // Use a jump table at the end, too.  If we did the copy_end
+        // overlapping store trick, we'd risk damaging the capability in
+        // the cell behind us.
+        small_copies<sizeof(void*), sizeof(long)>(dst, src, len);
+      }
+      // Otherwise, we cannot use pointer-width operations because one of
+      // the load or store is going to be misaligned.  So, same dance, but
+      // with integer registers only.
```

Comment:
Why not do a vectorised `memcpy` here?  On Morello, doing a Neon 128-bit memcpy is probably faster than doing a 64-bit one.